clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -50,7 +55,8 @@
 public class Fetcher<K, V> extends AbstractFetch<K, V> {
 
     private final Logger log;
-    private final AtomicBoolean isClosed = new AtomicBoolean(false);
+    private final FetchCollector<K, V> fetchCollector;
+    private final ConsumerNetworkClient client;
```

Comment:
I agree, it's ugly 😦

`client` and `nodeStatusDetector` both point at the same object in `Fetcher` and `AbstractFetch`, respectively, but the two classes use them for different purposes.

The `ConsumerNetworkClient` used in the `Fetcher` invokes methods like `poll()`, `send()`, and `disableWakeups()` to perform network I/O "directly." The forthcoming `FetchRequestManager` adheres to the design principle that request managers don't perform network I/O directly. To enforce that, we introduced this intentionally limited `NodeStatusDetector` to allow for checking basic state of nodes.

I'll keep looking at it a bit to see if I can come up with something cleaner because I don't like it either.

Reply:
I made a change here that I'm still not crazy about.

Instead of having a `NodeStatusDetector` instance variable in `AbstractFetch`, I introduced a type parameter in `AbstractFetch` that allows a subclass to provide a more specific type. Now `Fetcher` can provide the type as `ConsumerNetworkClient` and thus it can call the specialized methods on the instance variable from `AbstractFetch`.

The part I'm not crazy about is that the instance variable is still called `nodeStatusDetector`. I think I would be a bit confused when I see calls like `nodeStatusDetector.disableWakeups()`. But maybe that's just me. I thought of changing the name of the instance variable to `client` or `networkInfo` or other equally vague names.

If you have any input/suggestions/preferences, do LMK. Thanks!